das/daser.go

```diff
@@ -285,51 +176,60 @@ func (d *DASer) sampleHeader(ctx context.Context, h *header.ExtendedHeader) erro
 	return nil
 }
 
-// SampleRoutineState reports the current state of the
-// DASer's main sampling routine.
-func (d *DASer) SampleRoutineState() RoutineState {
-	d.state.sampleLk.RLock()
-	state := d.state.sample
-	d.state.sampleLk.RUnlock()
-	return state
-}
+func (d *DASer) runSubscriber(ctx context.Context, sub header.Subscription, emit listenFn) {
+	defer sub.Cancel()
 
-func (d *DASer) updateSampleState(h *header.ExtendedHeader, err error) {
-	height := uint64(h.Height)
+	for {
+		// subscribe for new headers to keep sampling process up-to-date with current network state
+		h, err := sub.NextHeader(ctx)
+		if err != nil {
+			if err == context.Canceled {
+				close(d.subscriberDone)
+				return
+			}
 
-	d.state.sampleLk.Lock()
-	defer d.state.sampleLk.Unlock()
-	d.state.sample.LatestSampledHeight = height
-	d.state.sample.LatestSampledSquareWidth = uint64(len(h.DAH.RowsRoots))
-	d.state.sample.Error = err
-}
+			log.Errorw("failed to get next header", "Err", err)
+			continue
+		}
+		log.Infow("found new header", "height", h.Height)
 
-func (d *DASer) indicateRunning() {
-	d.state.sampleLk.Lock()
-	defer d.state.sampleLk.Unlock()
-	d.state.sample.IsRunning = true
+		emit(ctx, uint64(h.Height))
```

Comment:
Ideally, we would avoid "regetting" the header and use the one we have, but that's not a big deal in this case

Reply:
Do you refer to this lines?
```
func (s *coordinatorState) updateMaxKnown(last uint64) {
	// seen this header before
	if last <= s.maxKnown {
		return
	}
	...
```
This case is added just in case GetHeader() API changes behavior. If it happens DASer will be screwed without this case.

I've added log WARN to track this specific scenario

Reply:
No, I meant the subscription gives the actual Header while we just ignore it and then get it again via HeaderGetter.

Reply:
Ah, I see. It means, there are will be extra call for header store for new headers. Not ideal, but should be fast considering headers cache

Reply:
Yes